File: leetcode121.go
// +build ignore

package main

import "fmt"

func main() {
	fmt.Print(maxProfit([]int{7, 6, 4, 3, 1}))
}

func maxProfit(prices []int) int {
	min := -1
	best := 0
	n := len(prices)
	for i := 0; i < n; i++ {
		if min == -1 || prices[i] < min {
			min = prices[i]
		}
		delta := prices[i] - min
		if delta > best {
			best = delta
		}
	}
	return best
}

File: leetcode121.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)
        best = 0
        for i in range(n-1, -1, -1):
            if i == 0:
                break
            for j in range(i-1, -1, -1):
                if prices[i] - prices[j] > best:
                    best = prices[i] - prices[j]
        return best
a = Solution()
print(a.maxProfit([7,1,5,3,6,4]))
print(a.maxProfit([7,6,4,3,1]))